### PR TITLE
SPU: Rework reverb buffer logic

### LINKED ISSUE #6769: [BUG]: 18 Wheeler: American Pro Trucker Audio Bug (Regression)
### Describe the Bug

I don't know if this bug is related to #6737 or not, but it seems somewhat similar. The audio periodically cuts out, or will play a super high pitch noise, and sometimes the character voices will shift way up in pitch. This didn't happen in 1.6.0. I would actually say the issue is more severe here than it is in Summoner (if it's the same issue) rendering the game basically unplayable.

### Reproduction Steps

Start the game, sit on the character selection for a brief amount of time or start playing and play briefly and the bug should appear.

### Expected Behavior

For the audio to play normally, like it did in 1.6.0.

### PCSX2 Revision

v1.7.3153

### Operating System

Windows 10 (64bit)

### If Linux - Specify Distro

_No response_

### CPU

Ryzen 9 5900X

### GPU

RTX 3080

### GS Settings

_No response_

### Emulation Settings

_No response_

### GS Window Screenshots

_No response_

### Logs & Dumps

_No response_

### PR DESCRIPTION
**Pull request by Ziemas**
Should fix 18 wheeler and the suffering.

Fixes #6769 

### Suggested Testing Steps

Make sure the music keeps playing in 18 wheeler and that reverb is consistent in the suffering intro.

And that nothing else in the vast PS2 library broke? :^)

### CODE REVIEW

**Comment by Ziemas** [comment]
needs more work to make 18 wheeler and the suffering both happy at the same time.

**Comment by Ziemas** [comment]
Did some rework instead, not ready to be merged.

**Comment by seta-san** [comment]
This might be an issue to look at

https://github.com/PCSX2/pcsx2/issues/2093

**Comment by Ziemas** [comment]
I think this works now? I have one more thing I want to HW test before deciding on this though.

**Comment by refractionpcsx2** [comment]
Yep, both games seem to work, nice job :) not sure how the mask of fff ffff works though, the reg only apparently has 3f as top bits available :/

**Comment by Ziemas** [comment]
Actually, 0x3f'ffff will work as well. I just needed some of the top bits to not matter (and can't really be arsed testing my way to the answer).

I hadn't actually looked at the manual, I'll just go with that instead then.

**Comment by Ziemas** [comment]
This should work fine, it doesn't handle the weird edge cases I found but I don't think any game hits those.

**Review by kamfretoz — APPROVED** [review]

File: pcsx2/SPU2/Reverb.cpp [hunk 1]
```diff
@@ -140,14 +142,24 @@ StereoOut32 __forceinline V_Core::ReverbUpsample(bool phase)
...
+__forceinline s32 V_Core::RevbGetIndexer(s32 offset)
+{
+	u32 start = EffectsStartA & 0x3f'ffff;
+	u32 end = (EffectsEndA & 0x3f'ffff) | 0xffff;
+
+	u32 x = ((Cycles >> 1) + offset) % ((end - start) + 1);
+
+	x += start;
+
+	return x & 0xf'ffff;
```
Comment (refractionpcsx2) [inline pcsx2/SPU2/Reverb.cpp:157]: Shouldn't that be 3fffff?

Comment (Ziemas) [reply pcsx2/SPU2/Reverb.cpp:157]: It's the final mask before directly indexing into _spu2mem so i want to mask it to fit within the memory size.

Comment (refractionpcsx2) [reply pcsx2/SPU2/Reverb.cpp:157]: fair enough :)

**Review by Mrlinkwii — APPROVED** [review]

**Pull request merged**